fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java

```diff
@@ -33,22 +36,73 @@
 	/**
 	 * Determines the mining level of the passed stack, which is used for calculating what blocks this tool is allowed to break.
 	 *
-	 * @param stack The stack to check on.
-	 * @param user The current user of the tool, or null if there isn't any.
+	 * @param stack The stack mining the block
+	 * @param user  The current user of the tool, or null if there isn't any
 	 * @return The mining level of the item. 3 is equal to a diamond pick.
+	 * @deprecated Use {@link #getMiningLevel(Tag, BlockState, ItemStack, LivingEntity)} to detect tag and block.
 	 */
-	//TODO: nullable on user once we have an official @Nullable annotation in
-	int getMiningLevel(ItemStack stack, LivingEntity user);
+	// nullable on user once we have an official @Nullable annotation in
+	@Deprecated
+	default int getMiningLevel(ItemStack stack, LivingEntity user) {
+		return 0;
+	}
+
+	/**
+	 * Determines the mining level of the passed stack, which is used for calculating what blocks this tool is allowed to break.
+	 *
+	 * @param tag   The tool tag the item stack is being compared to
+	 * @param state The block to mine
+	 * @param stack The stack mining the block
+	 * @param user  The current user of the tool, or null if there isn't any
+	 * @return The mining level of the item. 3 is equal to a diamond pick.
+	 */
+	// nullable on user once we have an official @Nullable annotation in
+	default int getMiningLevel(Tag<Item> tag, BlockState state, ItemStack stack, LivingEntity user) {
```

Comment:
What is nullable here?

The user?

If it's the user I would change to this for now:
```suggestion
	default int getMiningLevel(Tag<Item> tag, BlockState state, ItemStack stack, /* @Nullable */ LivingEntity user) {
```